### PR TITLE
HCIDOCS-187: Document procedure for changing BMC address after installation

### PR DESCRIPTION
**Pull request by johnwilkins**
Added a module for editing a BareMetalHost resource after installation.

Fixes: [HCIDOCS-187](https://issues.redhat.com//browse/HCIDOCS-187)

See https://issues.redhat.com/browse/HCIDOCS-187 for additional details.

Preview URL: https://76624--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_bare_metal_ipi/ipi-install-installation-workflow.html#bmc-addressing_ipi-install-installation-workflow and https://76624--ocpdocs-pr.netlify.app/openshift-enterprise/latest/post_installation_configuration/bare-metal-configuration.html#editing-a-baremetalhost-resource_post-install-bare-metal-configuration

For release(s): 4.16
QE Review: 

- [x] QE has approved this change. 

Signed-off-by: John Wilkins <jowilkin@redhat.com>

### CODE REVIEW

File: modules/ipi-install-bmc-addressing.adoc [hunk 1]
```diff
@@ -8,6 +8,8 @@
 
 Most vendors support Baseboard Management Controller (BMC) addressing with the Intelligent Platform Management Interface (IPMI). IPMI does not encrypt communications. It is suitable for use within a data center over a secured or dedicated management network. Check with your vendor to see if they support Redfish network boot. Redfish delivers simple and secure management for converged, hybrid IT and the Software Defined Data Center (SDDC). Redfish is human readable and machine capable, and leverages common internet and web services standards to expose information directly to the modern tool chain. If your hardware does not support Redfish network boot, use IPMI.
 
+If you need to modify the BMC address or credentials after installation, see the _Additional resources_ section.
```
Comment (MahnoorAsghar) [inline modules/ipi-install-bmc-addressing.adoc:11]: This epic applies to BMC address only, not credentials. While your statement is correct in the context of this doc, the following must be made clear about editing the BMC address specifically:

The BMC address can be edited in two ways/cases: with the detached annotation, and when the BMH is still in the 'Registering' state (in BMO). 
[From my observation, a BMH very quickly moves out of the Registering state as soon as the credentials are verified, but if you enter the BMC address incorrectly (by mistake, for example), the BMH will not move past the Registering state.] While the BMH is in the Registering state, the BMC address can be edited in the BMH resource (without the need for detachment). In all other BMH states, you have to use the detach annotation.

Comment (johnwilkins) [reply modules/ipi-install-bmc-addressing.adoc]: @MahnoorAsghar I've made some changes. Can you have a look and give an LGTM if it's okay?

File: modules/bmo-editing-a-baremetalhost-resource.adoc [hunk 2]
```diff
@@ -0,0 +1,46 @@
+// This module is included in the following assemblies: 
+//
+// post_installation_configuration/bare-metal-configuration.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="editing-a-baremetalhost-resource_{context}"]
+= Editing a BareMetalHost resource
+
+After deploying a cluster, you might need to edit a node's `BareMetalHost` resource. For example, if you deploy a cluster with the {ai-full} and you need to add baseboard management controller credentials, or you want to move a node from one cluster to another without deprovisioning it, you must edit the node's `BareMetalHost` resource. To edit a node's `BareMetalHost` resource without deprovisioning the node, use the following procedure.
```
Comment (dfitzmau) [inline modules/bmo-editing-a-baremetalhost-resource.adoc:9]: baseboard management controller > baseboard management controller (BMC)

Comment (dfitzmau) [reply modules/bmo-editing-a-baremetalhost-resource.adoc:9]: I see this acronym used in our docs.

File: modules/bmo-editing-a-baremetalhost-resource.adoc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+----
+<1> Replace `<node_name>` with the name of the node.
+
+. Edit the  `BareMetalHost` resource by running the following command:
++
+[source,terminal]
+----
+$ oc edit bmh <node_name> -n openshift-machine-api <1>
+----
+<1> Replace `<node_name>` with the name of the node.
```
Comment (dfitzmau) [inline modules/bmo-editing-a-baremetalhost-resource.adoc:38]: I think the last 2 of these callouts can be deleted as the first example gives enough indication.

File: modules/bmo-editing-a-baremetalhost-resource.adoc [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+. Finally, reattach the node to Ironic by running the following command:
```
Comment (dfitzmau) [inline modules/bmo-editing-a-baremetalhost-resource.adoc:40]: The "Finally" transitional phrase does not add much value. Step 4 is the last step so no further indication is needed.

File: modules/bmo-editing-a-baremetalhost-resource.adoc [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
++
+[source,terminal]
+----
+$ oc annotate baremetalhost <node_name> -n openshift-machine-api 'baremetalhost.metal3.io/detached'- <1>
```
Comment (dfitzmau) [inline modules/bmo-editing-a-baremetalhost-resource.adoc:44]: Highlighting on GH but teh rendered content look OK in the preview, maybe add the following to all example blocks?

[source,terminal,subs="+quotes"]

File: modules/ipi-install-bmc-addressing.adoc [hunk 6]
```diff
@@ -8,6 +8,8 @@
 
 Most vendors support Baseboard Management Controller (BMC) addressing with the Intelligent Platform Management Interface (IPMI). IPMI does not encrypt communications. It is suitable for use within a data center over a secured or dedicated management network. Check with your vendor to see if they support Redfish network boot. Redfish delivers simple and secure management for converged, hybrid IT and the Software Defined Data Center (SDDC). Redfish is human readable and machine capable, and leverages common internet and web services standards to expose information directly to the modern tool chain. If your hardware does not support Redfish network boot, use IPMI.
 
+You can modify the BMC address during installation while the node is in the `Registering` state. If you need to modify the BMC address after the node leaves the `Registering` state, or if you need to modify the BMC credentials, you must disconnect the node from Ironic, edit the `BareMetalHost` resource, and reconnect the node to Ironic. See the _Additional resources_ section for details.
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:11]: Is there a way of creating a unique add resources under this section as the one I think your pointing to is very far down the document. If someone adds an add resource elsewhere, confusion coukld arise.

File: modules/ipi-install-bmc-addressing.adoc [hunk 7]
```diff
@@ -64,59 +66,99 @@ platform:
...
+----
+<1> Replace `<username>` with the name of the BMC user.
+
+. Set the BMC password by running the following command:
++
+[source,terminal]
+----
+$ export PASS=<password> <1>
+----
+<1> Replace `<password>` with the password of the BMC user.
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:96]: I'm not srue these are needed. Customers should be familiar with use-repalced values. If we had a large example, calling out replaceable values would be cumbersome if loads existed.

File: modules/ipi-install-bmc-addressing.adoc [hunk 8]
```diff
@@ -64,59 +66,99 @@ platform:
...
+----
+<1> Replace `<ip_address>` with the IP address of the BMC.
+
+. Set the ID of the system by running the following command:
++
+[source,terminal]
+----
+$ export SystemID=<system_id> <1>
+----
+<1> Replace `<system_id>` with the system ID. For example, `System.Embedded.1` or `1`. See the following vendor-specific BMC sections for details.
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:112]: This callout is needed as it provides useful input details.

File: modules/ipi-install-bmc-addressing.adoc [hunk 9]
```diff
@@ -64,59 +66,99 @@ platform:
...
+
+* Check `power off` support by running the following command:
 +
 [source,terminal]
 ----
-curl -u $USER:$PASS -X POST -H'Content-Type: application/json' -H'Accept: application/json' -d '{"ResetType": "ForceOff"}' https://$SERVER/redfish/v1/Systems/$SystemID/Actions/ComputerSystem.Reset
+$ curl -u $USER:$PASS -X POST -H'Content-Type: application/json' -H'Accept: application/json' -d '{"ResetType": "ForceOff"}' https://$SERVER/redfish/v1/Systems/$SystemID/Actions/ComputerSystem.Reset
 ----
-* Temporary boot using `pxe`
+* Check temporary boot using `pxe` by running the following command:
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:129]:
```suggestion
* Check the temporary boot implementation that uses `pxe` by running the following command:
```

File: modules/ipi-install-bmc-addressing.adoc [hunk 10]
```diff
@@ -64,59 +66,99 @@ platform:
...
 +
 [source,terminal]
 ----
-curl -u $USER:$PASS -X PATCH -H "Content-Type: application/json"  https://$Server/redfish/v1/Systems/$SystemID/ -d '{"Boot": {"BootSourceOverrideTarget": "pxe", "BootSourceOverrideEnabled": "Once"}}
+$ curl -u $USER:$PASS -X PATCH -H "Content-Type: application/json"  https://$Server/redfish/v1/Systems/$SystemID/ -d '{"Boot": {"BootSourceOverrideTarget": "pxe", "BootSourceOverrideEnabled": "Once"}}
 ----
-* Set BIOS boot mode using `Legacy` or `UEFI`
+
+* Check the ability to set the BIOS boot mode using `Legacy` or `UEFI` by running the following command:
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:136]:
```suggestion
* Check the status of setting the BIOS boot mode that uses `Legacy` or `UEFI` by running the following command:
```

File: modules/ipi-install-bmc-addressing.adoc [hunk 11]
```diff
@@ -64,59 +66,99 @@ platform:
...
 ----
-curl -u $USER:$PASS -X PATCH -H "Content-Type: application/json"  https://$Server/redfish/v1/Systems/$SystemID/ -d '{"Boot": {"BootSourceOverrideMode":"UEFI"}}
+$ curl -u $USER:$PASS -X PATCH -H "Content-Type: application/json"  https://$Server/redfish/v1/Systems/$SystemID/ -d '{"Boot": {"BootSourceOverrideMode":"UEFI"}}
 ----
 
-List of redfish-virtualmedia APIs::
-* Set temporary boot device using `cd` or `dvd`
+.List of Redfish virtual media APIs:
+
+* Check the ability to set the temporary boot device using `cd` or `dvd` by running the following command:
```
Comment (dfitzmau) [inline modules/ipi-install-bmc-addressing.adoc:145]:
```suggestion
* Check the ability to set the temporary boot device that uses `cd` or `dvd` by running the following command:
```

**Comment by dfitzmau** [comment]
Hi @johnwilkins . Nice work. Stylistic comments added.

/remove-label peer-review-in-progress

/label peer-review-done

File: modules/ipi-install-bmc-addressing.adoc (hunk 6 above)
Comment (MahnoorAsghar) [inline modules/ipi-install-bmc-addressing.adoc:11]: I highly suspect that you do not need to detach the BMH for editing the BMC credentials.

File: modules/ipi-install-bmc-addressing.adoc (hunk 8 above)
Comment (MahnoorAsghar) [reply modules/ipi-install-bmc-addressing.adoc:112]: Agreed.

File: modules/ipi-install-bmc-addressing.adoc (hunk 7 above)
Comment (MahnoorAsghar) [reply modules/ipi-install-bmc-addressing.adoc:96]: (Not from the docs team, but FWIW I agree, its unnecessary and increases doc length)

**Comment by MahnoorAsghar** [comment]
While the stylistic improvements are a good improvement, they are not directly related to the HCIDOCS-187 card (even though they touch adjacent files). It makes it easier to review if unrelated improvements are made in a separate PR (even if they refer to the same card). (2 reviews can be done, no worries about that).

File: modules/ipi-install-bmc-addressing.adoc (hunk 6 above)
Comment (zaneb) [reply modules/ipi-install-bmc-addressing.adoc:11]: Yes, @MahnoorAsghar is correct. You can change the password any time.
To change the _address_ - effectively pointing the BMH at a completely different host - you must now disconnect from ironic first (previously this was not allowed at all).

File: modules/bmo-editing-a-baremetalhost-resource.adoc (hunk 2 above)
Comment (MahnoorAsghar) [reply modules/bmo-editing-a-baremetalhost-resource.adoc:9]: Editing a BMH's credentials and moving a BMH from one cluster to another are two vastly different scenarios. The first does not require the BMH to be detached. Editing BMC address, on the other hand, requires either the BMH to be in the Registering state, or to be detached.

File: modules/bmo-editing-a-baremetalhost-resource.adoc [hunk 12]
```diff
@@ -0,0 +1,44 @@
+// This module is included in the following assemblies: 
+//
+// post_installation_configuration/bare-metal-configuration.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="editing-a-baremetalhost-resource_{context}"]
+= Editing a BareMetalHost resource
+
+After deploying a cluster, you might need to edit a node's `BareMetalHost` resource. For example, if you deploy a cluster with the {ai-full} and you need to add the baseboard management controller (BMC) host name or IP address, or you want to move a node from one cluster to another without deprovisioning it, you must edit the node's `BareMetalHost` resource. To edit a node's `BareMetalHost` resource without deprovisioning the node, use the following procedure.
```
Comment (MahnoorAsghar) [inline modules/bmo-editing-a-baremetalhost-resource.adoc:9]: 'BMC address' should be the term used. (BMC) Hostname can cause confusion with a system hostname, which is an entirely different thing. BMC address is clear and self-explanatory.

Comment (johnwilkins) [reply modules/bmo-editing-a-baremetalhost-resource.adoc:9]: I can change it if you want, but in my home setup, I can address the BMC with either an IP address or its host name. BMCs can have a host name in my experience and testing.

Comment (MahnoorAsghar) [reply modules/bmo-editing-a-baremetalhost-resource.adoc:9]: Would that not be the domain name, similar to google.com or github.com?

**Pull request closed without merging**